core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/search/LexicalSimilarityTermFilter.java

```diff
@@ -0,0 +1,222 @@
+		this.prefixLength = prefixLength;
+		this.maxExpansions = maxExpansions;
+	}
+	
+	public String getTerm() {
+		return term;
+	}
+	
+	public Integer getMinShouldMatch() {
+		return minShouldMatch;
+	}
+	
+	public Boolean isIgnoreStopwords() {
+		return ignoreStopwords;
+	}
+	
+	public boolean isCaseSensitive() {
+		return caseSensitive;
+	}
+	
+	public Boolean isSynonyms() {
+		return synonyms;
+	}
+	
+	public String getFuzziness() {
+		return fuzziness;
+	}
+	
+	public Integer getPrefixLength() {
+		return prefixLength;
+	}
+	
+	public Integer getMaxExpansions() {
+		return maxExpansions;
+	}
+	
+	@JsonIgnore
+	public boolean isAnyMatch() {
+		return getMinShouldMatch() != null;
+	}
+	
+	public LexicalSimilarityTermFilter withIgnoreStopwords() {
+		return new Builder(this).ignoreStopwords(true).build();
+	}
+	
+	public LexicalSimilarityTermFilter withTerm(String newTerm) {
+		return new Builder(this).term(newTerm).build();
+	}
+	
+	@Override
+	public Set<String> getTerms() {
+		return Set.of(term);
+	}
+	
+	@Override
+	public Expression toExpression(String field, String textFieldSuffix, String exactFieldSuffix, String prefixFieldSuffix) {
+		// lexical similarity based on the following queries
+		return dismaxWithScoreCategories(
+			// exact match overrides anything and should be scored the highest
+			TermFilter.exact().term(getTerm()).caseSensitive(isCaseSensitive()).build().toExpression(field, textFieldSuffix, exactFieldSuffix, prefixFieldSuffix),
+
+			// matching based on synonyms and tokenized text, less score than exact but more than the other (??? not sure for fuzziness though)
+			matchTextAll(fieldAlias(field, textFieldSuffix), getTerm())
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms()),
+			matchBooleanPrefix(fieldAlias(field, textFieldSuffix), getTerm())
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms()),
+
+			// matching based on fuzziness should receive higher scores than leaving out words
+			matchTextAll(fieldAlias(field, textFieldSuffix), getTerm()).withFuzziness(fuzziness, prefixLength, maxExpansions),
+			
+			// leaving out words but still matching some completely should generate a better similarity than prefix only
+			matchTextAny(fieldAlias(field, textFieldSuffix), getTerm(), getMinShouldMatch())
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms()),
+
+			// then all prefixes match
+			matchTextAll(fieldAlias(field, prefixFieldSuffix), getTerm())
+				.withIgnoreStopwords(isIgnoreStopwords()),
+				
+			// and last some prefixes match
+			matchTextAny(fieldAlias(field, prefixFieldSuffix), getTerm(), getMinShouldMatch())
+				.withIgnoreStopwords(isIgnoreStopwords())
+				.withSynonymsEnabled(isSynonyms())
+		);
+	}
+	
+	public static final class Builder {
+		
+		private String term;
+		private Integer minShouldMatch;
+		
+		private Boolean ignoreStopwords;
+		private boolean caseSensitive;
```

Comment:
Is `caseSensitive` a mandatory option? The other two fields are `Boolean` wrappers, not primitives.